src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStep.java

```diff
@@ -176,92 +176,95 @@ private void doStart() throws IOException, InterruptedException {
+            @Override
+            public void onSuccess(StepContext context, Object result) {
+                try {
+                    finished(context, true);
+                } catch (Exception x) {
+                    context.onFailure(x);
+                    return;
+                }
+                context.onSuccess(result);
             }
-            context.onFailure(t);
-        }
 
-        private void finished(StepContext context, boolean success) throws Exception {
-            Run<?, ?> r = context.get(Run.class);
-            TestResult provisional = null;
-            for (PipelineRealtimeTestResultAction a : r.getActions(PipelineRealtimeTestResultAction.class)) {
-                if (a.id.equals(id)) {
-                    provisional = a.getResult();
-                    r.removeAction(a);
-                    LOGGER.log(Level.FINE, "clearing {0} from {1}", new Object[] {id, r});
-                    AbstractRealtimeTestResultAction.saveBuild(r);
-                    break;
+            @Override
+            public void onFailure(StepContext context, Throwable t) {
+                try {
+                    finished(context, false);
+                } catch (Exception x) {
+                    t.addSuppressed(x);
                 }
+                context.onFailure(t);
             }
-            if (!success) {
-                archiver.setAllowEmptyResults(true);
-            }
-            TaskListener listener = context.get(TaskListener.class);
-            try {
-                FilePath workspace = context.get(FilePath.class);
-                workspace.mkdirs();
-                Launcher launcher = context.get(Launcher.class);
-                FlowNode node = context.get(FlowNode.class);
-
-                List<FlowNode> enclosingBlocks = JUnitResultsStepExecution.getEnclosingStagesAndParallels(node);
-
-                PipelineTestDetails pipelineTestDetails = new PipelineTestDetails();
-                pipelineTestDetails.setNodeId(id);
-                pipelineTestDetails.setEnclosingBlocks(JUnitResultsStepExecution.getEnclosingBlockIds(enclosingBlocks));
-                pipelineTestDetails.setEnclosingBlockNames(JUnitResultsStepExecution.getEnclosingBlockNames(enclosingBlocks));
-
-                TestResultSummary summary = JUnitResultArchiver.parseAndSummarize(archiver, pipelineTestDetails, r, workspace, launcher, listener);
-
-                if (summary.getFailCount() > 0) {
-                    int testFailures = summary.getFailCount();
-                    if (testFailures > 0) {
-                        node.addOrReplaceAction(new WarningAction(Result.UNSTABLE).withMessage(testFailures + " tests failed"));
-                        if (!archiver.isSkipMarkingBuildUnstable()) {
-                            r.setResult(Result.UNSTABLE);
+
+            private void finished(StepContext context, boolean success) throws Exception {
+                run(() -> {
```

Comment:
This means that `context.onSuccess/onFailure` will run _before_ the work below. Is that really what you wanted? Compare https://github.com/jenkinsci/workflow-step-api-plugin/blob/b8e7c15bc95a090629e78d446c478b80c0d49c92/src/main/java/org/jenkinsci/plugins/workflow/steps/GeneralNonBlockingStepExecution.java#L137-L162

Reply:
Ah thanks, pushed, un tested but the tests pass for it